### PR TITLE
skip converting empty yaml objects to json

### PR DESCRIPTION
**Pull request by ashwgit**
PTAL, this one skips empty YAML documents during FBC YAML→JSON conversion to prevent emitting null objects that breaks opm validate while performing add operations. 
Also fixes a bug for including conversion for .yml file too.

### CODE REVIEW

**Comment by ashwgit** [comment]
@yashvardhannanavati PTAL

File: tests/test_workers/test_tasks/test_fbc_utils.py [hunk 1]
```diff
@@ -300,3 +300,25 @@ def test__serialize_datetime():
 def test__serialize_datetime_raise():
     with pytest.raises(TypeError, match="Type <class 'int'> is not serializable."):
         _serialize_datetime(2025)
+
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply tests/test_workers/test_tasks/test_fbc_utils.py:303]: added

File: iib/workers/tasks/fbc_utils.py [hunk 2]
```diff
@@ -159,7 +159,10 @@ def enforce_json_config_dir(config_dir: str) -> None:
     for dirpath, _, filenames in os.walk(config_dir):
         for file in filenames:
             in_file = os.path.join(dirpath, file)
-            if in_file.lower().endswith(".yaml"):
+            if in_file.lower().endswith(".yaml") or in_file.lower().endswith(".yml"):
+                if os.path.getsize(in_file) == 0:
+                    raise ValidationError(f"Empty YAML file found: {in_file}")
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py]: done

File: iib/workers/tasks/fbc_utils.py [hunk 3]
```diff
@@ -159,7 +159,10 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
+
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:166]: done

File: iib/workers/tasks/fbc_utils.py [hunk 4]
```diff
@@ -159,7 +159,10 @@ def enforce_json_config_dir(config_dir: str) -> None:
     for dirpath, _, filenames in os.walk(config_dir):
         for file in filenames:
             in_file = os.path.join(dirpath, file)
-            if in_file.lower().endswith(".yaml"):
+            if in_file.lower().endswith(".yaml") or in_file.lower().endswith(".yml"):
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:163]: I don't want to increase the clod path code for such corner cases.

File: tests/test_workers/test_tasks/test_fbc_utils.py [hunk 5]
```diff
@@ -300,3 +300,25 @@ def test__serialize_datetime():
...
+    foo: bar
+    ---
+    ---
+    another: data
+    """
+
+    expected_result = '{"foo": "bar"}{"another": "data"}'
+
+    input = os.path.join(tmpdir, "test_file.yaml")
+    output = os.path.join(tmpdir, "test_file.json")
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply tests/test_workers/test_tasks/test_fbc_utils.py]: patched

File: iib/workers/tasks/fbc_utils.py [hunk 6]
```diff
@@ -169,5 +172,8 @@ def enforce_json_config_dir(config_dir: str) -> None:
                 with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
                     data = yaml.load_all(yaml_in)
                     for chunk in data:
+                        # Ignore if it is an empty yaml object
```
[the remark this answers is not in the document]
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:176]: Not Required

File: iib/workers/tasks/fbc_utils.py [hunk 7]
```diff
@@ -154,12 +154,16 @@ def enforce_json_config_dir(config_dir: str) -> None:
...
 
     :param str config_dir: The config dir to walk recursively converting any YAML to JSON.
+    :raises IIBError: If the yaml content is empty.
     """
     log.info("Enforcing JSON content on config_dir: %s", config_dir)
     for dirpath, _, filenames in os.walk(config_dir):
         for file in filenames:
             in_file = os.path.join(dirpath, file)
-            if in_file.lower().endswith(".yaml"):
+            if in_file.lower().endswith(".yaml") or in_file.lower().endswith(".yml"):
```
Comment (MichalZelenak) [inline iib/workers/tasks/fbc_utils.py:163]: As we previously did not support the "yml" - I would throw rather throw an IIBError here, because as fullsend pointed out - there could be a case of foo.yaml and foo.yml -> causing silent data loss (first processed file would vanish).

File: iib/workers/tasks/fbc_utils.py [hunk 8]
```diff
@@ -169,5 +173,8 @@ def enforce_json_config_dir(config_dir: str) -> None:
                 with open(in_file, 'r') as yaml_in, open(out_file, 'a') as json_out:
                     data = yaml.load_all(yaml_in)
                     for chunk in data:
+                        # Ignore if it is an empty yaml object
+                        if chunk is None:
+                            continue
```
Comment (MichalZelenak) [inline iib/workers/tasks/fbc_utils.py:178]: Yaml like this (empty one):

[empty_docs.yaml](https://github.com/user-attachments/files/30343731/empty_docs.yaml)

would correctly remove the empty yaml file, but the loop bellow would leave out an empty json file - issue would persist.

Comment (MichalZelenak) [reply iib/workers/tasks/fbc_utils.py:178]: Could you also add a test with YAML like this one?

Comment (MichalZelenak) [inline iib/workers/tasks/fbc_utils.py:178]: This seems like a fix candidate:

```
  wrote_any = False
  for chunk in data:
      if chunk is None:
          continue
      json.dump(chunk, json_out, default=_serialize_datetime)
      wrote_any = True
  if not wrote_any:
      # I am not sure if we should throw an error here, or just remove the empty yaml silently
     raise IIBError(f"YAML file contains no valid documents: {in_file}")
```

Comment (MichalZelenak) [reply iib/workers/tasks/fbc_utils.py:178]: Even better solution IMO (Quick draft):

```
  # load the full YAML in
  with open(in_file, 'r') as yaml_in:
      chunks = [chunk for chunk in yaml.load_all(yaml_in) if chunk is
  not None]

  # write the data, if there is something
  if chunks:
      with contextlib.suppress(FileNotFoundError):
          os.remove(out_file)
      with open(out_file, 'w') as json_out:
          for chunk in chunks:
              json.dump(chunk, json_out, default=_serialize_datetime)
      os.remove(in_file)
```

**Review by MichalZelenak — CHANGES_REQUESTED** [review]
Left a few comments, please take a look.

**Comment by ashwgit** [comment]
I will again open this PR as I had to delete me forked repo. I will patch the requested changes over there.

**Comment by ashwgit** [comment]
https://github.com/release-engineering/iib/pull/1372

File: iib/workers/tasks/fbc_utils.py (hunk 7 above)
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:163]: @chandwanitulsi Do we expect this edge case ? 

Note- Any new changes will be introduced to #1372

File: iib/workers/tasks/fbc_utils.py (hunk 8 above)
Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:178]: @MichalZelenak added here - #1372 on line https://github.com/release-engineering/iib/pull/1372/changes#diff-4883d86ddcc636748009a842e2eea1416fb70ba4ee6ccc609f1219f8b9446dabR182

Comment (ashwgit) [reply iib/workers/tasks/fbc_utils.py:178]: Thanks a lot, I have implemented it.
